src/common/viz/ConfigDialog.js

```diff
@@ -20,36 +22,54 @@ define([
         DESCRIPTION_BASE = $('<div class="desc muted col-sm-8"></div>'),
         SECTION_HEADER = $('<h6 class="config-section-header">');
 
+    const CUSTOM_WIDGETS = JSON.parse(CustomWidgets);
+
     var ConfigDialog = function(client) {
         PluginConfigDialog.call(this, {client: client});
         this._widgets = {};
+        this.imported = this._registerCustomWidgets(CUSTOM_WIDGETS);
     };
 
     ConfigDialog.prototype = Object.create(PluginConfigDialog.prototype);
 
+    ConfigDialog.prototype._registerCustomWidgets = async function(customWidgets) {
+        const self = this;
+        const promises = customWidgets.map(widgetInfo => {
+            return new Promise((resolve, reject) => {
+                require([widgetInfo.path], function(customWidget) {
+                    const targetType = widgetInfo.type;
+                    self._propertyGridWidgetManager
+                        .registerWidgetForType(targetType, customWidget);
+                    resolve();
+                }, reject);
+            });
+        });
+        return Promise.all(promises);
+    };
+
     ConfigDialog.prototype.show = async function(pluginMetadata, options={}) {
         const deferred = Q.defer();
-
         this._pluginMetadata = pluginMetadata;
         const prevConfig = await this.getSavedConfig();
+        await this.imported;
         this._initDialog(pluginMetadata, prevConfig, options);
 
         this._dialog.on('shown', () => {
             this._dialog.find('input').first().focus();
         });
 
-        this._btnSave.on('click', event => {
-            this.submit(deferred.resolve);
+        this._btnSave.on('click', async event => {
+            await this.submit(deferred.resolve);
```

Comment:
hmm.. Does jquery support async event handlers correctly here? Maybe we should move `event.stopPropagation()` and `event.preventDefault()` to be called first.

The issue I am concerned about is that I am not sure that the underlying implementation is something like this:
```javascript
let target = clickedItem;
let handled = false;
while (!handled) {
    target.onclick(event);
    target = target.parent();
    handled = event.propagate || !!target;  // assuming `stopPropagation` toggles a flag, `propagate`
}
```
In the above case, the event would still propagate until the async function completes.